internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -50,71 +54,60 @@ func (k *Kubeadm) PreFunc(data *userdata.UserData) (err error) {
-			CertPath string
-			KeyPath  string
-		}{
-			{
-				Cert:     data.Security.Kubernetes.CA,
-				CertPath: constants.KubeadmCACert,
-				KeyPath:  constants.KubeadmCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.SA,
-				CertPath: constants.KubeadmSACert,
-				KeyPath:  constants.KubeadmSAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.FrontProxy,
-				CertPath: constants.KubeadmFrontProxyCACert,
-				KeyPath:  constants.KubeadmFrontProxyCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.Etcd,
-				CertPath: constants.KubeadmEtcdCACert,
-				KeyPath:  constants.KubeadmEtcdCAKey,
-			},
-		}
-
-		for _, cert := range certs {
-			if cert.Cert == nil {
-				continue
-			}
-			if err = writeKubeadmPKIFiles(cert.Cert, cert.CertPath, cert.KeyPath); err != nil {
-				return err
-			}
-		}
-	} else if data.Services.Kubeadm.IsControlPlane() {
-		if data.Services.Trustd == nil || data.Services.Trustd.BootstrapNode == "" {
-			return nil
-		}
+	// Write out all certs we've been provided
+	if err = writeKubeadmPKIFiles(data); err != nil {
+		return err
+	}
 
-		creds, err := basic.NewCredentials(data.Services.Trustd)
-		if err != nil {
+	// Run kubeadm init phase certs all. This should fill in whatever gaps
+	// we have in the provided certs.
+	if data.Services.Kubeadm.IsBootstrap() {
+		cmd := exec.Command(
+			"kubeadm",
+			"init",
+			"phase",
+			"certs",
+			"all",
+			"--config",
+			constants.KubeadmConfig)
+		if err = cmd.Run(); err != nil {
 			return err
 		}
+	}
 
-		files := []string{
-			constants.AuditPolicyPathInitramfs,
-			constants.EncryptionConfigInitramfsPath,
-		}
+	if data.Services.Kubeadm.IsWorker() || data.Services.Kubeadm.IsBootstrap() || data.Services.Trustd == nil {
+		log.Println("Skipping retrieval of files from peers via trustd")
+		return nil
+	}
 
-		conn, err := basic.NewConnection(data.Services.Trustd.BootstrapNode, constants.TrustdPort, creds)
-		if err != nil {
-			return err
-		}
+	// Get a list of files we need to request
+	// ( filtered by ones we already have )
+	filelist := fileSet()
 
-		client := proto.NewTrustdClient(conn)
+	// Initialize trustd peer client connection
+	var trustds []proto.TrustdClient
+	if trustds, err = createTrustdClients(data); err != nil {
+		return err
+	}
 
-		if err := writeFiles(client, files); err != nil {
-			return err
+	// Get assets from remote nodes
+	for _, fileRequest := range filelist {
+		// Have a single chan shared across all clients
+		// for a given file
+		content := make(chan []byte, len(trustds))
+		donezo := make(chan struct{})
+
+		// kick off a goroutine for each trustd client
+		// to fetch the given file
+		for _, trustdClient := range trustds {
+			go download(trustdClient, fileRequest, content, donezo)
 		}
-
+		// read from the content chan to write out the
+		// given file
+		go writeTrustdFiles(fileRequest.Path, content, donezo)
 	}
 
+	// No need to block since `ConditionFunc` will block for us
```

Comment:
there might be some good things in running this more explicitly, like how do we want to report progress? failures? what if all the trustd endpoints are down, but there's no way to find that, as service would be simply 'waiting for files'?

Reply:
in other words, if `pre` state returns an error (say, it can't write some file, that is not retryable), we can see the error via `osctl service`, if it returns `nil` and relies on `wait` state, we can 't do anything

Reply:
Thats where we say tough bananas 🍌 

Would an err chan to be appropriate to add in to address that?

Reply:
I think it could have blocked execution until it's done. We can add `context` to `PreFunc` so that it can abort if node is shutdown. But otherwise service will be in 'pre' state until it fetches all the certs. I have a PR which simplifies that to some extent - 'pre' runs after 'condition', so that we run 'pre' when we're ready (dependent services have already started, etc.)